Makes Zip Guns Buildable

SiegDerMaus opened the pull request:
<!-- 
Do not forget to add a changelog when you have made admin/player facing changes that can alter gameplay.
Examples which require a changelog entry include:
* Adding/removing objects that players may interact with, or the way they function.
* Adding/removing/altering admin tools.
* Changing the map.

Examples were changelog entries are optional/not typically required:
* Cosmetic changes such as descriptions, sound effects, etc.
* Optimizations and other changes to underlying systems which do not affect gameplay.
* Minor bug fixes.

You find a README and example file in .\html\changelogs\ for further instructions.
-->

Allows zip guns to be built in game from a wooden stock, atmos pipe, tape, and a mouse trap. Also gives them their own sprite.

SiegDerMaus committed:
Makes Zip Guns Buildable

Allows zip guns to be built in game. Also gives them their own sprite.

Techhead0 commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
+		update_icon()
+		return
+	else if(istype(thing,/obj/item/weapon/tape_roll) && buildstate == 1)
+		user.visible_message("<span class='notice'>\The [user] secures \the assembly with \the [thing].</span>")
```
`secures \the assembly` should just be `secures the assembly`. You don't need the text macro here. (See also other cases below.)

Techhead0 commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
 	..()
+
+// Zip gun construction.
+/obj/item/weapon/gun/projectile/zipgunframe
```
Making `zipgunframe` a child of `/obj/item/weapon/gun/projectile` might not be the best idea. It's not a gun yet, and shouldn't inherit gun-like behaviour. `/obj/item/weapon/zipgunframe` might be better.

GinjaNinja32 commented on code/modules/projectiles/guns/projectile.dm:
```diff
@@ -36,13 +36,14 @@
-			loaded += new ammo_type(src)
-	if(ispath(magazine_type) && (load_method & MAGAZINE))
-		ammo_magazine = new magazine_type(src)
+	if (starts_loaded == 1)
```
no need for ==1 here, just `if(starts_loaded)`.

Techhead0 commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
+		qdel(thing)
+		user.visible_message("<span class='notice'>\The [user] fits \the [thing] to the \the [src] as a crude barrel.</span>")
+		buildstate++
+		update_icon()
```
Consider:
```
/obj/item/weapon/gun/projectile/zipgunframe/proc/advance_buildstate()
	add_fingerprint(user)
	buildstate++
	update_icon()
```
Then you can insert said proc into each intermediate step, and `transfer_fingerprints_to(zipgun)` in the last step.

GinjaNinja32 commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
+
+/obj/item/weapon/gun/projectile/zipgunframe/attackby(var/obj/item/thing, var/mob/user)
+	if(istype(thing,/obj/item/pipe) && buildstate == 0)
+		user.drop_from_inventory(thing)
```
Won't this allow making a barrel with... for example....

- a bent pipe?
- a 4-way pipe?
- a gas pump?
- a scrubber?

SiegDerMaus replied:
I never thought of that. How do I rectify that?

GinjaNinja32 replied:
Not sure.

SiegDerMaus replied:
All of those objects would have straight segments of pipes as components. Couldn't someone conceivably break those segments off?

Techhead0 replied:
I just looked, and pneumatic cannons also can be built with any `/obj/item/pipe`.

SiegDerMaus replied:
So it's a larger problem, then?

BlueNexus replied:
Suggestion, check 
```
if(thing.type == /obj/pathstuff/etc)
```
Might work.

SiegDerMaus replied:
Since it's a larger problem with other craftable objects that use pipes, isn't it a bit beyond the scope of this? It can be fixed later, of course.

Techhead0 reviewed (commented):
In additions to the line-comments, I don't see any way to actually make the zip gun frame itself.

SiegDerMaus commented:
Oh, right, I forgot one of the files. I'll make those changes.

GinjaNinja32 commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
+			put_in_hands = M == user
+			M.drop_from_inventory(src)
+		if(put_in_hands)
+			user.put_in_hands(zipgun)
```
Cleaner:
```dm
if(ismob(loc))
    var/mob/M = loc
    M.drop_from_inventory(src)
    M.put_in_hands(zipgun)
```

If the object is in someone's inventory, there's no way a mob other than that mob can attack it with a screwdriver. Just assume loc == user if loc is a mob.

SiegDerMaus committed:
Fixes and Clean Up

Cleans up the code a bit and makes zip gun frame a child of
/obj/item/weapon to fix it having gun behaviour.

SiegDerMaus committed:
Fixes Text Macros

This should finally fix all the text macros. Hopefully.

PsiOmegaDelta commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -181,10 +181,65 @@
+	icon_state = "zipgun[buildstate]"
+
+/obj/item/weapon/zipgunframe/examine(mob/user)
+	..(user)
```
. = ..()
The return value may be of importance.

PsiOmegaDelta commented on code/modules/projectiles/guns/projectile.dm:
```diff
@@ -36,13 +36,14 @@
 	//var/list/ammo_states = list()	//values
 
-/obj/item/weapon/gun/projectile/New()
+/obj/item/weapon/gun/projectile/New(atom/loc, starts_loaded = 1)
```
Hmm.. while this isn't wrong as such the current standard is to have a var on the object type that's overriden, rather than a proc argument. Guess I need to decide if the change is needed.

PsiOmegaDelta replied:
How to relocate this to a var instead:
* Remove the New() argument
* Add the variable to the type:
```
/obj/item/weapon/gun/projectile
	var/starts_loaded = TRUE
````
* Override it for any types that should begin unloaded:
```
/obj/item/weapon/gun/projectile/not_pre_loaded
	starts_loaded = FALSE
```
* This value can be overwritten pre-New() using the { } syntax if needed.

SiegDerMaus committed:
Makes Psi's Suggested Changes

SiegDerMaus committed:
Fixes a Silly

Fixes a silly oversight. Something left in the code from a previous
version that was unneeded.

SiegDerMaus commented:
@GinjaNinja32 and @PsiOmegaDelta ? Can you guys take another look at this to see if I've made your changes to a satisfactory degree?
